### PR TITLE
#3 Added vector implementation

### PR DESCRIPTION
**Pull request by titova-olga**

### CODE REVIEW

File: src/Vector.cpp [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+    /*non default copyable*/
+    Vector(const IVector& other) = delete;
+    void operator=(const Vector& other) = delete;
+};
+}
+
+Vector::Vector(unsigned int size, const double *vals)
+{
+    this->size = size;
+    this->vals = new(std::nothrow) double[size];
```
Comment (MaximSmolskiy) [inline src/Vector.cpp]: It is necessary to add a check that the memory was allocated

File: src/Vector.cpp [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+    if (!right)
+    {
+        ILog::report("IVector.add: Can't add vector, right - nullptr.\n");
+        return ERR_WRONG_ARG;
+    }
+    DIM_CHECK(right, "add");
+
+    int errCode;
+    double coord;
+    double *vals = new double[size];
```
Comment (MaximSmolskiy) [inline src/Vector.cpp]: Please add std::nothrow

File: src/Vector.cpp [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+    if (!vals)
+    {
+        ILog::report("IVector.add: Not enough memory.\n");
+        return NULL;
```
Comment (MaximSmolskiy) [inline src/Vector.cpp]: Please replace by ERR_MEMORY_ALLOCATION

File: src/Vector.cpp [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+    if (!right)
+    {
+        ILog::report("IVector.subtract: Can't subtract vector, right - nullptr.\n");
+        return ERR_WRONG_ARG;
+    }
+    DIM_CHECK(right, "subtract");
+
+    int errCode;
+    double coord;
+    double *vals = new double[size];
```
Comment (MaximSmolskiy) [inline src/Vector.cpp]: Please add std::nothrow

File: src/Vector.cpp [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+    if (!vals)
+    {
+        ILog::report("IVector.add: Not enough memory.\n");
+        return NULL;
```
Comment (MaximSmolskiy) [inline src/Vector.cpp]: Please replace by ERR_MEMORY_ALLOCATION

File: src/Vector.cpp [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+
+IVector* IVector::add(IVector const* const left, IVector const* const right)
+{
+    if (!right || !left || left->getDim() != right->getDim())
+    {
+        ILog::report("IVector.add: Can't add vector, some of vectors - nullptr or dimensions mismatch.\n");
+        return NULL;
+    }
+
+    IVector *v = left->clone();
```
Comment (MaximSmolskiy) [inline src/Vector.cpp:219]: Please add check v != NULL

File: src/Vector.cpp [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+
+IVector* IVector::subtract(IVector const* const left, IVector const* const right)
+{
+    if (!right || !left || left->getDim() != right->getDim())
+    {
+        ILog::report("IVector.subtract: Can't subtract vector, some of vectors - nullptr or dimensions mismatch.\n");
+        return NULL;
+    }
+
+    IVector *v = left->clone();
```
Comment (MaximSmolskiy) [inline src/Vector.cpp:243]: Please add check v != NULL

File: src/Vector.cpp [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+
+IVector* IVector::multiplyByScalar(IVector const* const left, double scalar)
+{
+    if (!left)
+    {
+        ILog::report("IVector.multiplyByScalar: Can't multiply vector by scalar, vector - nullpt.\n");
+        return NULL;
+    }
+
+    IVector *v = left->clone();
```
Comment (MaximSmolskiy) [inline src/Vector.cpp:267]: Please add check v != NULL

File: src/Vector.cpp [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+        ILog::report("IVector.eq: Can't subtract vectors.\n");
+        return ERR_WRONG_ARG;
+    }
+
+    double norm;
+    int errCode = v->norm(type, norm);
+    if (errCode != ERR_OK)
+        return errCode;
+
+    result = abs(norm) < precision;
```
Comment (MaximSmolskiy) [inline src/Vector.cpp]: "abs" no need here

File: src/Vector.cpp [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+
+    delete[] v;
```
Comment (MaximSmolskiy) [inline src/Vector.cpp]: It should be "delete v;" (without [])

File: src/Vector.cpp [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+    if (!v)
+    {
+        ILog::report("IVector.eq: Can't subtract vectors.\n");
+        return ERR_WRONG_ARG;
+    }
+
+    double norm;
+    int errCode = v->norm(type, norm);
+    if (errCode != ERR_OK)
+        return errCode;
```
Comment (maik-nack) [inline src/Vector.cpp:433]: It needs to delete `v` in this case. You can delete this after 
```c++
    int errCode = v->norm(type, norm);
```
and erase 
```c++
    delete v;
```
on line 428.

File: src/Vector.cpp [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+        return ERR_WRONG_ARG;
+    }
+    DIM_CHECK(right, "subtract");
+
+    int errCode;
+    double coord;
+    double *vals = new(std::nothrow) double[size];
+    if (!vals)
+    {
+        ILog::report("IVector.add: Not enough memory.\n");
```
Comment (maik-nack) [inline src/Vector.cpp]: It is not `IVector.add`. It is `IVector.subtract`.

File: src/Vector.cpp [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+public:
+
+    int getId() const;
+
+    /*operations*/
+    int add(IVector const* const right);
+    int subtract(IVector const* const right);
+    int multiplyByScalar(double scalar);
+    int dotProduct(IVector const* const right, double& res) const;
+    int crossProduct(IVector const* const right);
```
Comment (maik-nack) [inline src/Vector.cpp]: It doesn't need an override.

File: src/Vector.cpp [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+        if (errCode != ERR_OK)
+            return errCode;
+
+        res += vals[i] * coord;
+    }
+
+    return ERR_OK;
+}
+
+int Vector::crossProduct(IVector const* const right)
```
Comment (maik-nack) [inline src/Vector.cpp]: It doesn't need an override.

File: src/Vector.cpp [hunk 15]
```diff
@@ -0,0 +1,436 @@
...
+        return NULL;
+    }
+
+    IVector *v = left->clone();
+    if (!v)
+    {
+        ILog::report("IVector.add: Can't add vector, can't clone vector.\n");
+        return NULL;
+    }
+    v->add(right);
```
Comment (maik-nack) [inline src/Vector.cpp]: It needs to check the error code from the `add` function.

File: src/Vector.cpp [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+        return NULL;
+    }
+
+    IVector *v = left->clone();
+    if (!v)
+    {
+        ILog::report("IVector.subtract: Can't subtract vector, can't clone vector.\n");
+        return NULL;
+    }
+    v->subtract(right);
```
Comment (maik-nack) [inline src/Vector.cpp]: It needs to check the error code from the `subtract` function.

File: src/Vector.cpp [hunk 17]
```diff
@@ -0,0 +1,436 @@
...
+        return NULL;
+    }
+
+    IVector *v = left->clone();
+    if (!v)
+    {
+        ILog::report("IVector.multiplyByScalar: Can't multiply vector by scalar, can't clone vector.\n");
+        return NULL;
+    }
+    v->multiplyByScalar(scalar);
```
Comment (maik-nack) [inline src/Vector.cpp]: It needs to check the error code from the `multiplyByScalar` function.

File: src/Vector.cpp [hunk 18]
```diff
@@ -0,0 +1,436 @@
...
+    result = norm < precision;
+
+    delete v;
+
+    return ERR_OK;
+}
+
+int Vector::getId() const
+{
+    return INTERFACE_0;
```
Comment (maik-nack) [inline src/Vector.cpp]: It may look better if you write `IVector::INTERFACE_0`. What do you think?

**Review by maik-nack — APPROVED** [review]
GREAT JOB!!!

**Review by MaximSmolskiy — APPROVED** [review]
Looks great!

**Pull request merged**